src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4286,384 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+			// Mouse button is down and mouse pointer entered a new inventory field
+
+			if (!s.isValid() || s.listname == "craftpreview")
+				break;
+
+			if (!m_selected_item && event.MouseInput.Shift) {
+				// Shift-move items while dragging
+				if (m_held_button == BET_RIGHT)
+					shift_move_amount = 1;
+				else if (m_held_button == BET_MIDDLE)
+					shift_move_amount = MYMIN(s_count, 10);
+				else if (m_held_button == BET_LEFT)
+					shift_move_amount = s_count;
+
+			} else if (m_selected_item) {
+				if (m_held_button != BET_LEFT) {
+					// Move items if the destination slot is empty
+					// or contains the same item type as what is going to be moved
+					if (!m_selected_dragging && (empty || matching)) {
+						if (m_held_button == BET_RIGHT)
+							move_amount = 1;
+						else if (m_held_button == BET_MIDDLE)
+							move_amount = MYMIN(m_selected_amount, 10);
+					}
+
+				} else if (m_left_dragging && (empty || matching) &&
+						   m_left_drag_amount > m_left_drag_stacks.size()) {
+					// Add the slot to the left-drag list if it doesn't exist
+					bool found = false;
+					for (auto &ds : m_left_drag_stacks) {
+						if (s.inventoryloc == ds.first->inventoryloc &&
+								s.listname == ds.first->listname && s.i == ds.first->i) {
+							found = true;
+						}
+					}
+					if (!found) {
+						m_left_drag_stacks[new GUIInventoryList::ItemSpec(s)] =
+								list_s->getItem(s.i);
+					}
+
+				} else if (m_selected_dragging && matching && !identical) {
+					// Pickup items of the same type while dragging
+					pickup_amount = s_count;
 				}
+
+			} else if (m_held_button == BET_LEFT) {
+				// Start picking up items
+				m_selected_item = new GUIInventoryList::ItemSpec(s);
+				m_selected_amount = s_count;
+				m_selected_dragging = true;
 			}
-		break;
+			break;
+		}
+		case BET_OTHER: {
+			// Some other mouse event has occured
+			// Currently only left-double-click should trigger this
+			if (!s.isValid() || event.MouseInput.Event != EMIE_LMOUSE_DOUBLE_CLICK)
+				break;
+
+			// Both the selected item and the hovered item need to be checked
+			// because we don't know exactly when the double-click happened
+			ItemStack slct;
+			if (!m_selected_item && !empty)
+				slct = list_s->getItem(s.i);
+			else if (m_selected_item && (identical || empty))
+				slct = list_selected->getItem(m_selected_item->i);
+
+			// Pickup all of the item from the list
+			if (slct.count > 0) {
+				for (u32 i = 0; i < list_s->getSize(); i++) {
+					// Skip the selected slot
+					if (i == s.i)
+						continue;
+
+					ItemStack item = list_s->getItem(i);
+
+					if (slct.name == item.name && slct.wear == item.wear &&
+							slct.metadata == item.metadata) {
+						// Found a match, check if we can pick it up
+						bool full = false;
+						u16 amount = item.count;
+						ItemStack leftover = slct.addItem(item, m_client->idef());
+						if (!leftover.empty()) {
+							amount -= leftover.count;
+							full = true;
+						}
```

Comment:
`item.itemFits(slct, OUTPUT, m_client->idef())` should be shorter and more reliable.

Reply:
See other comment regarding this above.